File: src/CompilerDispatcher/CompilerDispatcher.java
package CompilerDispatcher;

import CHTL.CHTLLexer.CHTLLexer;
import CHTL.CHTLLexer.Token;
import CHTL.CHTLLoader.CHTLLoader;
import CHTL.CHTLManage.DefinitionManager;
import CHTL.CHTLParser.CHTLParser;
import CHTL.CHTLNode.BaseNode;
import CHTL.CHTLProcessor.ProcessedAST;
import CHTL.CHTLProcessor.ASTProcessor;
import CHTL.CHTLGenerator.CHTLGenerator;

import java.util.List;

public class CompilerDispatcher {
    public static void main(String[] args) {
        // A CHTL source code example for testing the import system.
        String source = "[Import] @Chtl from \"other.chtl\"\n" +
                "\n" +
                "body {\n" +
                "    @Element ImportedBox from other;\n" +
                "}";

        System.out.println("--- CHTL Source ---");
        System.out.println(source);

        try {
            // 0. Setup
            DefinitionManager definitionManager = new DefinitionManager();
            CHTLLoader loader = new CHTLLoader();

            // 1. Lexer
            CHTLLexer lexer = new CHTLLexer(source);
            List<Token> tokens = lexer.tokenize();

            // 2. Parser
            // The initial parser is for the main file, in the default namespace.
            CHTLParser parser = new CHTLParser(tokens, definitionManager, DefinitionManager.DEFAULT_NAMESPACE, loader);
            List<BaseNode> ast = parser.parse();

            // 3. AST Processor
            ASTProcessor astProcessor = new ASTProcessor(definitionManager);
            ProcessedAST processedAST = astProcessor.process(ast);

            // 4. Generator
            CHTLGenerator generator = new CHTLGenerator();
            String bodyContent = generator.generate(processedAST.ast());
            String styleBlock = generator.generateStyleBlock(processedAST.globalStyles());

            // 5. Final Assembly
            String finalHtml;
            if (!styleBlock.isEmpty() && bodyContent.contains("</head>")) {
                finalHtml = bodyContent.replace("</head>", styleBlock + "</head>");
            } else {
                finalHtml = bodyContent;
            }

            System.out.println("\n--- Generated HTML ---");
            System.out.println(finalHtml);

        } catch (Exception e) {
            System.err.println("\n--- Compilation Failed ---");
            e.printStackTrace();
        }
    }
}

File: src/CompilerDispatcher/CompilerDispatcher.cpp
#include "CompilerDispatcher.h"
#include "../CHTL/CHTLLexer.h"
#include "../CHTL/CHTLParser.h"
#include "../CHTL/CHTLGenerator.h"
#include "../Scanner/UnifiedScanner.h"
#include "../CodeMerger/CodeMerger.h"

namespace CHTL {

CompilerDispatcher::CompilerDispatcher() 
    : inlineCSS_(false), inlineJS_(false), minify_(false), debugMode_(false) {
}

CompileResult CompilerDispatcher::compile(const std::string& source) {
    CompileResult result;
    
    try {
        // 使用统一扫描器分离代码片段
        UnifiedScanner scanner;
        ScanResult scanResult = scanner.scan(source);
        
        // 词法分析
        CHTLLexer lexer(source);
        TokenList tokens = lexer.tokenize();
        
        // 语法分析
        CHTLParser parser(tokens);
        NodePtr root = parser.parse();
        
        if (!root) {
            result.errors.push_back("Failed to parse source");
            return result;
        }
        
        // 代码生成
        CHTLGenerator generator;
        GenerationOptions options;
        options.inlineCSS = inlineCSS_;
        options.inlineJS = inlineJS_;
        options.minify = minify_;
        options.debugMode = debugMode_;
        
        GenerationResult genResult = generator.generate(root, options);
        
        if (genResult.hasErrors()) {
            result.errors.insert(result.errors.end(), genResult.errors.begin(), genResult.errors.end());
        }
        
        if (genResult.hasWarnings()) {
            result.warnings.insert(result.warnings.end(), genResult.warnings.begin(), genResult.warnings.end());
        }
        
        result.html = genResult.html;
        result.css = genResult.css;
        result.js = genResult.js;
        
    } catch (const std::exception& e) {
        result.errors.push_back("Compilation error: " + std::string(e.what()));
    }
    
    return result;
}

CompileResult CompilerDispatcher::compileFile(const std::string& filename) {
    CompileResult result;
    
    try {
        std::ifstream file(filename);
        if (!file.is_open()) {
            result.errors.push_back("Cannot open file: " + filename);
            return result;
        }
        
        std::string source((std::istreambuf_iterator<char>(file)),
                          std::istreambuf_iterator<char>());
        file.close();
        
        return compile(source);
        
    } catch (const std::exception& e) {
        result.errors.push_back("File compilation error: " + std::string(e.what()));
    }
    
    return result;
}

void CompilerDispatcher::setErrorHandler(std::function<void(const std::string&)> handler) {
    errorHandler_ = handler;
}

void CompilerDispatcher::reportError(const std::string& message) {
    if (errorHandler_) {
        errorHandler_(message);
    }
}

} // namespace CHTL